lumicks/pylake/force_calibration/detail/hydrodynamics.py

```diff
@@ -0,0 +1,193 @@
+    include it however. We decided to follow the paper and omit the second term. The effects of
+    this difference were very small for all simulations tested.
+        m_star = bead_mass + (2.0 * np.pi * rho_sample * radius ** 3.0) / 3.0"""
+    bead_mass = (4.0 / 3.0) * np.pi * bead_radius ** 3 * rho_bead
+    return gamma0 / (2.0 * np.pi * bead_mass)
+
+
+def calculate_complex_drag(f, gamma0, rho_sample, bead_radius, distance_to_surface=None):
+    """Calculates the complex drag coefficient which takes into account surface effects
+    analogously to [1].
+
+    Parameters
+    ----------
+    f : float
+        Frequency [Hz]
+    gamma0 : float
+        Drag coefficient [mPas]
+    bead_radius : float
+        Bead radius [m]
+    rho_sample : float
+        Sample mass density [kg/m^3]
+    distance_to_surface : float
+        Distance from bead center to nearest surface [m]
+
+    [1] Tolić-Nørrelykke S. F, Schäffer E, Howard J, Pavone F. S, Jülicher F and Flyvbjerg, H.
+    Calibration of optical tweezers with positional detection in the back focal plane, Review of
+    scientific instruments 77, 103101 (2006).
+    """
+    # frequency_nu is the frequency at which the penetration depth in the liquid of the
+    # bead’s linear harmonic motion equals the radius of the bead. It parametrizes the
+    # flow pattern established around a sphere undergoing linear harmonic oscillations in
+    # an incompressible fluid and it is unrelated to the mass of particle.
+    #
+    # The equation given in [6] is:
+    #   nu / (pi * radius**2)
+    #
+    # Where nu is the kinematic viscosity. In the paper gamma0 is defined as:
+    #
+    #   gamma0 = 6 * pi * rho_sample * nu * radius
+    #
+    # Hence, we can obtain nu (kinematic viscosity) as:
+    #
+    #   gamma0 / (6 * pi * rho_sample * radius)
+    nu = gamma0 / (6.0 * np.pi * rho_sample * bead_radius)
+    frequency_nu = nu / (np.pi * bead_radius ** 2)
+
+    freq_ratio = f / frequency_nu
+    sqrt_freq_ratio = np.sqrt(freq_ratio)
+
+    # Stokes contribution to the drag coefficient (gamma / gamma0 from the paper)
+    # Equation D4 from [1]
+    real_stokes = 1 + sqrt_freq_ratio
+    imag_stokes = -sqrt_freq_ratio - 2 / 9 * freq_ratio
+
+    # If we have no distance specified, we assume we are very far from the surface, in which case
+    # Stokes drag applies.
+    if distance_to_surface is None:
+        return real_stokes, imag_stokes
+
+    epsilon = (2 * distance_to_surface - bead_radius) * sqrt_freq_ratio / bead_radius
+    r_over_l = bead_radius / distance_to_surface
+
+    # Effect of the surface (depends on depth) e.g. denominator from Eq D6 in [1].
+    # fmt: off
+    real_surf = 1 - 9 / 16 * r_over_l * (1 - sqrt_freq_ratio / 3 - 4 / 3 * (1 - np.exp(-epsilon) * np.cos(epsilon)))
+    imag_surf = - 9 / 16 * r_over_l * (sqrt_freq_ratio / 3 + 2 / 9 * freq_ratio + 4 / 3 * np.exp(-epsilon) * np.sin(epsilon))
+    # fmt: on
```

Comment:
these forms of the equations are from Berg Eqn 34, which is indeed `[1]` in the `calibration_models.py` reference list.  Additionally, those equations already define `Re or Im {gamma/gamma_0}` so the division below isn't needed(?)

Is this intentional to be mixing equations from 2 different manuscripts, since everything else if from Tolic `[6]`?

Reply:
They're just the equations from Tolic rewritten to look like the ones in Berg.

Welp, here goes nothing:

Starting from the denominator in Tolic.

![image](https://user-images.githubusercontent.com/19836026/131877179-018e2403-2f81-4a51-9a85-4c57a4b79faa.png)

Note that the notation is pretty strange here. It almost looks like they're doing 2 * i * f / 9 * f_nu, but the 9 and f_nu are actually together part of the denominator like `(9*fv)`. If you don't consider them together, the model blows up.

1 - (9/16) * (R/l) * {1 - [(1 - i)/3] sqrt(f/fv) + 2 * i * f / (9 * fv) - (4/3) * (1 - exp(-(1-i) * (2*l-R)/delta))}

Let's group some terms:

1 - (9/16) * (R/l) * {1 - [(1 - i)/3] sqrt(f/fv) + (2 / 9) * i * (f / fv) - (4/3) * (1 - exp(-(1-i) * (2*l-R)/delta))}

Write c = (2 * l - R) / delta to not have to write down so much.

1 - (9/16) * (R/l) * {1 - [(1 - i)/3] sqrt(f/fv) + (2 / 9) * i * (f / fv) - (4/3) * (1 - exp(-(1-i) * c))}

Multiply out the terms in the exponential:

1 - (9/16) * (R/l) * {1 - [(1 - i)/3] sqrt(f/fv) + (2 / 9) * i * (f / fv) - (4/3) * (1 - exp(-c + i * c))}

Split the complex part from the exponential:

1 - (9/16) * (R/l) * {1 - [(1 - i)/3] sqrt(f/fv) + (2 / 9) * i * (f / fv) - (4/3) * (1 - exp(-c) * exp(i * c))}

Rewrite the complex exponential as a cosine and sine term:

1 - (9/16) * (R/l) * {1 - [(1 - i)/3] sqrt(f/fv) + (2 / 9) * i * (f / fv) - (4/3) * (1 - exp(-c) * (cos(c) + i * sin(c)))}

Multiply out the terms:

1 - (9/16) * (R/l) * {1 - [(1 - i)/3] sqrt(f/fv) + (2 / 9) * i * (f / fv) - (4/3) * (1 - exp(-c) * cos(c) - exp(-c) * i * sin(c))}

We have multiplied everything out now, so Real and Imag is simply a matter of collecting the real and imaginary parts.

Real = 1 - (9/16) * (R/l) * {1 - [(1)/3] sqrt(f/fv) + (2 / 9) * i * (f / fv) - (4/3) * (1 - exp(-c) * cos(c) - exp(-c) * i * sin(c))}

Real = 1 - (9/16) * (R/l) * {1 - (1 / 3) * sqrt(f/fv) - (4/3) * (1 - exp(-c) * cos(c))}

Imag = - (9/16) * (R/l) * {- [(-1)/3] sqrt(f/fv) + (2 / 9) * (f / fv) - (4/3) * (- exp(-c) * sin(c))}

Imag = - (9/16) * (R/l) * {(1/3) sqrt(f/fv) + (2 / 9) * (f / fv) + (4/3) * (exp(-c) * sin(c))}

Just to verify that the final term corresponds to the denominator in Tolic, you can run the following in `Sympy`:

```
import sympy as sym

R, l, f, fv, c = sym.symbols('R, l, f, fv, c', real=True, positive=True)

fulleq = 1 - sym.Rational(9, 16) * (R/l) * (1 - (sym.Rational(1,3) * (1 - sym.I)) * sym.sqrt(f/fv) + sym.Rational(2, 9) * sym.I * f / fv - sym.Rational(4, 3) * (1 - sym.exp(-(1 - sym.I) * c)))

manual_real = 1 - sym.Rational(9, 16) * (R/l) * (1 - sym.Rational(1,3) * sym.sqrt(f/fv) - sym.Rational(4, 3) * (1 - sym.exp(-c) * sym.cos(c)))
manual_imag = - sym.Rational(9, 16) * (R/l) * (sym.Rational(1,3) * sym.sqrt(f/fv) + sym.Rational(2, 9) * (f / fv) + sym.Rational(4, 3) * (sym.exp(-c) * sym.sin(c)))

print(sym.re(fulleq) - manual_real)
print(sym.im(fulleq) - manual_imag)
```

Reply:
thanks for the detailed derivation 🤣 double checked by hand just to triple verify. I wonder if it might be worthwhile putting a small comment, something on the order of `"D6 from [6] re-written with Euler's formula to approximate the form used in [1]"` just so any user that's looking through the code (or us in the future) don't have the same initial reaction I did.

Reply:
Yeah, fair point. I've added a comment about it.